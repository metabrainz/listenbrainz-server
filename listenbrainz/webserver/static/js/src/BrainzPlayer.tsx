import * as React from "react";
import {
  isEqual as _isEqual,
  isNil as _isNil,
  isString as _isString,
  get as _get,
  has as _has,
} from "lodash";
import * as _ from "lodash";
import PlaybackControls from "./PlaybackControls";
import APIService from "./APIService";
import SpotifyPlayer from "./SpotifyPlayer";
import YoutubePlayer from "./YoutubePlayer";
import SoundcloudPlayer from "./SoundcloudPlayer";

export type DataSourceType = {
  playListen: (listen: Listen | JSPFTrack) => void;
  togglePlay: () => void;
  seekToPositionMs: (msTimecode: number) => void;
};

export type DataSourceTypes = SpotifyPlayer | YoutubePlayer | SoundcloudPlayer;

export type DataSourceProps = {
  show: boolean;
  playerPaused: boolean;
  onPlayerPausedChange: (paused: boolean) => void;
  onProgressChange: (progressMs: number) => void;
  onDurationChange: (durationMs: number) => void;
  onTrackInfoChange: (title: string, artist?: string) => void;
  onTrackEnd: () => void;
  onTrackNotFound: () => void;
  handleError: (error: BrainzPlayerError, title?: string) => void;
  handleWarning: (message: string | JSX.Element, title?: string) => void;
  handleSuccess: (message: string | JSX.Element, title?: string) => void;
  onInvalidateDataSource: (
    dataSource?: DataSourceTypes,
    message?: string | JSX.Element
  ) => void;
};

type BrainzPlayerProps = {
  spotifyUser: SpotifyUser;
  direction: BrainzPlayDirection;
  onCurrentListenChange: (listen: Listen | JSPFTrack) => void;
  currentListen?: Listen | JSPFTrack;
  listens: Array<Listen | JSPFTrack>;
  newAlert: (
    alertType: AlertType,
    title: string,
    message: string | JSX.Element
  ) => void;
  apiService: APIService;
};

type BrainzPlayerState = {
  currentDataSourceIndex: number;
  currentTrackName: string;
  currentTrackArtist?: string;
  direction: BrainzPlayDirection;
  playerPaused: boolean;
  durationMs: number;
  progressMs: number;
  updateTime: number;
};

export default class BrainzPlayer extends React.Component<
  BrainzPlayerProps,
  BrainzPlayerState
> {
  spotifyPlayer?: React.RefObject<SpotifyPlayer>;
  youtubePlayer?: React.RefObject<YoutubePlayer>;
  soundcloudPlayer?: React.RefObject<SoundcloudPlayer>;
  dataSources: Array<React.RefObject<DataSourceTypes>> = [];

  // Since we don't want autoplay on our pages, we need a way to know
  // that a user clicked on the play/pause button for the first time
  // to start the playlist. Subsequent uses should toggle play/pause.
  firstRun: boolean = true;

  playerStateTimerID?: NodeJS.Timeout;

  constructor(props: BrainzPlayerProps) {
    super(props);

    this.spotifyPlayer = React.createRef<SpotifyPlayer>();
    if (SpotifyPlayer.hasPermissions(props.spotifyUser)) {
      this.dataSources.push(this.spotifyPlayer);
    }

    this.youtubePlayer = React.createRef<YoutubePlayer>();
    this.dataSources.push(this.youtubePlayer);

    this.soundcloudPlayer = React.createRef<SoundcloudPlayer>();
    this.dataSources.push(this.soundcloudPlayer);

    this.state = {
      currentDataSourceIndex: 0,
      currentTrackName: this.getCurrentTrackName(),
      currentTrackArtist: this.getCurrentTrackArtists(),
      direction: props.direction || "down",
      playerPaused: true,
      progressMs: 0,
      durationMs: 0,
      updateTime: performance.now(),
    };
  }

  isCurrentListen = (element: Listen | JSPFTrack): boolean => {
    const { currentListen } = this.props;
    if (_isNil(currentListen)) {
      return false;
    }
    if (_has(element, "identifier")) {
      // JSPF Track format
      return (element as JSPFTrack).id === (currentListen as JSPFTrack).id;
    }
    return _isEqual(element, currentListen);
  };

  playPreviousTrack = (): void => {
    this.playNextTrack(true);
  };

  playNextTrack = (invert: boolean = false): void => {
    const { listens } = this.props;
    const { direction } = this.state;

    if (listens.length === 0) {
      this.handleWarning(
        "You can try loading listens or refreshing the page",
        "No listens to play"
      );
      return;
    }

    const currentListenIndex = listens.findIndex(this.isCurrentListen);

    let nextListenIndex;
    if (currentListenIndex === -1) {
      nextListenIndex = direction === "up" ? listens.length - 1 : 0;
    } else if (direction === "up") {
      nextListenIndex =
        invert === true ? currentListenIndex + 1 : currentListenIndex - 1 || 0;
    } else if (direction === "down") {
      nextListenIndex =
        invert === true ? currentListenIndex - 1 || 0 : currentListenIndex + 1;
    } else {
      this.handleWarning("Please select a song to play", "Unrecognised state");
      return;
    }

    const nextListen = listens[nextListenIndex];
    if (!nextListen) {
      this.handleWarning(
        "You can try loading more listens or refreshing the page",
        "No more listens to play"
      );
      return;
    }
    this.playListen(nextListen);
  };

  handleError = (error: BrainzPlayerError, title?: string): void => {
    const { newAlert } = this.props;
    if (!error) {
      return;
    }
    newAlert(
      "danger",
      title || "Playback error",
      _isString(error)
        ? error
        : `${!_isNil(error.status) ? `Error ${error.status}:` : ""} ${
            error.message || error.statusText
          }`
    );
  };

  handleWarning = (message: string | JSX.Element, title?: string): void => {
    const { newAlert } = this.props;
    newAlert("warning", title || "Playback error", message);
  };

  handleSuccess = (message: string | JSX.Element, title?: string): void => {
    const { newAlert } = this.props;
    newAlert("success", title || "Success", message);
  };

  invalidateDataSource = (
    dataSource?: DataSourceTypes,
    message?: string | JSX.Element
  ): void => {
    let { currentDataSourceIndex: dataSourceIndex } = this.state;
    if (dataSource) {
      dataSourceIndex = this.dataSources.findIndex(
        (source) => source.current === dataSource
      );
    }
    if (dataSourceIndex >= 0) {
      if (message) {
        this.handleWarning(message, "Cannot play from this source");
      }
      this.dataSources.splice(dataSourceIndex, 1);
    }
  };

  playListen = (
    listen: Listen | JSPFTrack,
    datasourceIndex: number = 0
  ): void => {
    if (this.firstRun) {
      this.firstRun = false;
    }

    /** If available, retreive the service the listen was listened with */
    let selectedDatasourceIndex = this.getSourceIndexByListenData(listen);

    /** If no matching datasource was found, revert to the default bahaviour
     * (play from source 0 or if called from failedToFindTrack, try next source)
     */
    if (selectedDatasourceIndex === -1) {
      selectedDatasourceIndex = datasourceIndex;
    }

    const { onCurrentListenChange } = this.props;
    onCurrentListenChange(listen);

    this.setState({ currentDataSourceIndex: selectedDatasourceIndex }, () => {
      const { currentDataSourceIndex } = this.state;
      const dataSource =
        this.dataSources[currentDataSourceIndex] &&
        this.dataSources[currentDataSourceIndex].current;
      if (!dataSource) {
        this.invalidateDataSource();
        return;
      }

      dataSource.playListen(listen);
    });
  };

  getSourceIndexByListenData = (listen: Listen | JSPFTrack): number => {
    let selectedDatasourceIndex = -1;
    const listeningFrom = _get(
      listen,
      "track_metadata.additional_info.listening_from"
    );
    const originURL = _get(listen, "track_metadata.additional_info.origin_url");

    /** Spotify */
    if (
      listeningFrom === "spotify" ||
      _get(listen, "track_metadata.additional_info.spotify_id")
    ) {
      selectedDatasourceIndex = this.dataSources.findIndex(
        (ds) => ds.current instanceof SpotifyPlayer
      );
    }

    /** Youtube */
    if (
      listeningFrom === "youtube" ||
      /youtube\.com\/watch\?/.test(originURL)
    ) {
      selectedDatasourceIndex = this.dataSources.findIndex(
        (ds) => ds.current instanceof YoutubePlayer
      );
    }

    /** SoundCloud */
    if (listeningFrom === "soundcloud" || /soundcloud\.com/.test(originURL)) {
      selectedDatasourceIndex = this.dataSources.findIndex(
        (ds) => ds.current instanceof SoundcloudPlayer
      );
    }

    return selectedDatasourceIndex;
  };

  togglePlay = async (): Promise<void> => {
    try {
      const { currentDataSourceIndex } = this.state;
      const dataSource =
        this.dataSources[currentDataSourceIndex] &&
        this.dataSources[currentDataSourceIndex].current;
      if (!dataSource) {
        this.invalidateDataSource();
        return;
      }
      await dataSource.togglePlay();
    } catch (error) {
      this.handleError(error);
    }
  };

  getCurrentTrackName = (): string => {
    const { currentListen } = this.props;
    return _get(currentListen, "track_metadata.track_name", "");
  };

  getCurrentTrackArtists = (): string | undefined => {
    const { currentListen } = this.props;
    return _get(currentListen, "track_metadata.artist_name", "");
  };

  seekToPositionMs = (msTimecode: number): void => {
    const { currentDataSourceIndex } = this.state;
    const dataSource =
      this.dataSources[currentDataSourceIndex] &&
      this.dataSources[currentDataSourceIndex].current;
    if (!dataSource) {
      this.invalidateDataSource();
      return;
    }
    dataSource.seekToPositionMs(msTimecode);
    this.progressChange(msTimecode);
  };

  toggleDirection = (): void => {
    this.setState((prevState) => {
      const direction = prevState.direction === "down" ? "up" : "down";
      return { direction };
    });
  };

  /* Listeners for datasource events */

  failedToFindTrack = (): void => {
    const { currentListen } = this.props;
    if (!currentListen) {
      this.playNextTrack();
      return;
    }
    const { currentDataSourceIndex } = this.state;

    if (currentDataSourceIndex < this.dataSources.length - 1) {
      // Try playing the listen with the next dataSource
      this.playListen(currentListen, currentDataSourceIndex + 1);
    } else {
      this.handleWarning(
        "We couldn't find a matching song on any music service we tried",
        "Oh no !"
      );
      this.playNextTrack();
    }
  };

  playerPauseChange = (paused: boolean): void => {
    this.setState({ playerPaused: paused }, () => {
      if (paused) {
        this.stopPlayerStateTimer();
      } else {
        this.startPlayerStateTimer();
      }
    });
  };

  progressChange = (progressMs: number): void => {
    this.setState({ progressMs, updateTime: performance.now() });
  };

  durationChange = (durationMs: number): void => {
    this.setState({ durationMs }, this.startPlayerStateTimer);
  };

  trackInfoChange = (title: string, artist?: string): void => {
    this.setState({ currentTrackName: title, currentTrackArtist: artist });
  };

  /* Updating the progress bar without calling any API to check current player state */

  startPlayerStateTimer = (): void => {
    this.stopPlayerStateTimer();
    this.playerStateTimerID = setInterval(() => {
      this.getStatePosition();
    }, 200);
  };

  getStatePosition = (): void => {
    let newProgressMs: number;
    const { playerPaused, durationMs, progressMs, updateTime } = this.state;
    if (playerPaused) {
      newProgressMs = progressMs || 0;
    } else {
      const position = progressMs + (performance.now() - updateTime);
      newProgressMs = position > durationMs ? durationMs : position;
    }
    this.setState({ progressMs: newProgressMs, updateTime: performance.now() });
  };

  stopPlayerStateTimer = (): void => {
    if (this.playerStateTimerID) {
      clearInterval(this.playerStateTimerID);
    }
    this.playerStateTimerID = undefined;
  };

  render() {
    const {
      currentDataSourceIndex,
      currentTrackName,
      currentTrackArtist,
      playerPaused,
      direction,
      progressMs,
      durationMs,
    } = this.state;
    const { spotifyUser, apiService } = this.props;
    return (
      <div>
        <PlaybackControls
          playPreviousTrack={this.playPreviousTrack}
          playNextTrack={this.playNextTrack}
          togglePlay={this.firstRun ? this.playNextTrack : this.togglePlay}
          playerPaused={playerPaused}
          toggleDirection={this.toggleDirection}
          direction={direction}
          trackName={currentTrackName}
          artistName={currentTrackArtist}
          progressMs={progressMs}
          durationMs={durationMs}
          seekToPositionMs={this.seekToPositionMs}
        >
          <SpotifyPlayer
            show={
              this.dataSources[currentDataSourceIndex]?.current instanceof
              SpotifyPlayer
            }
            refreshSpotifyToken={apiService.refreshSpotifyToken}
            onInvalidateDataSource={this.invalidateDataSource}
            ref={this.spotifyPlayer}
            spotifyUser={spotifyUser}
            playerPaused={playerPaused}
            onPlayerPausedChange={this.playerPauseChange}
            onProgressChange={this.progressChange}
            onDurationChange={this.durationChange}
            onTrackInfoChange={this.trackInfoChange}
            onTrackEnd={this.playNextTrack}
            onTrackNotFound={this.failedToFindTrack}
            handleError={this.handleError}
            handleWarning={this.handleWarning}
            handleSuccess={this.handleSuccess}
          />
          <YoutubePlayer
            show={
              this.dataSources[currentDataSourceIndex]?.current instanceof
              YoutubePlayer
            }
            onInvalidateDataSource={this.invalidateDataSource}
            ref={this.youtubePlayer}
            playerPaused={playerPaused}
            onPlayerPausedChange={this.playerPauseChange}
            onProgressChange={this.progressChange}
            onDurationChange={this.durationChange}
            onTrackInfoChange={this.trackInfoChange}
            onTrackEnd={this.playNextTrack}
            onTrackNotFound={this.failedToFindTrack}
            handleError={this.handleError}
            handleWarning={this.handleWarning}
            handleSuccess={this.handleSuccess}
          />
          <SoundcloudPlayer
            show={
              this.dataSources[currentDataSourceIndex]?.current instanceof
              SoundcloudPlayer
            }
            onInvalidateDataSource={this.invalidateDataSource}
            ref={this.soundcloudPlayer}
            playerPaused={playerPaused}
            onPlayerPausedChange={this.playerPauseChange}
            onProgressChange={this.progressChange}
            onDurationChange={this.durationChange}
            onTrackInfoChange={this.trackInfoChange}
            onTrackEnd={this.playNextTrack}
            onTrackNotFound={this.failedToFindTrack}
            handleError={this.handleError}
            handleWarning={this.handleWarning}
            handleSuccess={this.handleSuccess}
          />
        </PlaybackControls>
      </div>
    );
  }
}
