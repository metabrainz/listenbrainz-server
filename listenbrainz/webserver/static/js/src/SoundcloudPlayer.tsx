import * as React from "react";
import { get as _get, throttle as _throttle } from "lodash";
import { DataSourceType, DataSourceProps } from "./BrainzPlayer";

require("../lib/soundcloud-player-api");

enum SoundCloudHTML5WidgetEvents {
  LOAD_PROGRESS = "loadProgress", // fired periodically while the sound is loading.
  PLAY_PROGRESS = "playProgress", // fired periodically while the sound is playing.
  PLAY = "play", // fired when the sound begins to play.
  PAUSE = "pause", // fired when the sound pauses.
  FINISH = "finish", // fired when the sound finishes.
  SEEK = "seek", // fired when the user seeks.
  READY = "ready", // fired when the widget has loaded its data and is ready to accept external calls.
  OPEN_SHARE_PANEL = "sharePanelOpened", // Fired when the user clicks the download button.
  CLICK_DOWNLOAD = "downloadClicked", // Fired when the user clicks the buy button.
  CLICK_BUY = "buyClicked", // Fired when the share panel is opened. This happens when the user clicks the "Share" button, and at the end of the last sound.
  ERROR = "error", // Fired when an error message is displayed.
}

interface SoundCloudHTML5Widget {
  play(): void;
  pause(): void;
  toggle(): void;
  seekTo(milliseconds: number): void;
  currentTime(): number;
  bind(eventName: string, handler: Function): void;
  unbind(eventName: string): void;
  load(url: string, options: any): void;
  setVolume(volume: number): void;
  // Getters
  getVolume(callback: Function): void;
  getDuration(callback: Function): void;
  getPosition(callback: Function): void;
  getSounds(callback: Function): void;
  getCurrentSound(callback: Function): void;
  getCurrentSoundIndex(callback: Function): void;
  isPaused(callback: Function): void;
  // Navigation (if multiple sounds loaded)
  next(): void;
  prev(): void;
  skip(soundIndex: number): void;
}

type ProgressEvent = {
  soundId: number;
  loadedProgress: number;
  currentPosition: number;
  relativePosition: number;
};

type SoundcloudPlayerState = {
  currentSoundId?: number;
};

export default class SoundcloudPlayer
  extends React.Component<DataSourceProps, SoundcloudPlayerState>
  implements DataSourceType {
  iFrameRef?: React.RefObject<HTMLIFrameElement>;
  soundcloudPlayer?: SoundCloudHTML5Widget;
  supportsSearch = false;
  retries = 0;
  // HTML widget options: https://developers.soundcloud.com/docs/api/html5-widget#parameters
  options = {
    auto_play: true,
    show_artwork: true,
    visual: true,
    buying: false,
    liking: false,
    download: false,
    sharing: false,
    show_comments: false,
    show_playcount: false,
    show_user: false,
    hide_related: true,
  };

  constructor(props: DataSourceProps) {
    super(props);
    this.state = { currentSoundId: undefined };
    this.iFrameRef = React.createRef();
  }

  componentDidMount = () => {
    const { onInvalidateDataSource } = this.props;
    if (!(window as any).SC) {
      onInvalidateDataSource(this, "Soundcloud JS API did not load properly.");
      // Fallback to uncontrolled iframe player?
      return;
    }
    if (!this.iFrameRef || !this.iFrameRef.current) {
      onInvalidateDataSource(this, "SoundCloud IFrame not found in page.");
      return;
    }

    this.soundcloudPlayer = (window as any).SC.Widget(
      this.iFrameRef.current
    ) as SoundCloudHTML5Widget;
    this.soundcloudPlayer.bind(
      SoundCloudHTML5WidgetEvents.READY,
      this.onReady.bind(this)
    );
  };

  componentDidUpdate(prevProps: DataSourceProps) {
    const { show } = this.props;
    if (prevProps.show === true && show === false && this.soundcloudPlayer) {
      this.soundcloudPlayer.pause();
    }
  }

  componentWillUnmount() {
    if (!this.soundcloudPlayer) {
      return;
    }
    this.soundcloudPlayer.unbind(SoundCloudHTML5WidgetEvents.FINISH);
    this.soundcloudPlayer.unbind(SoundCloudHTML5WidgetEvents.PAUSE);
    this.soundcloudPlayer.unbind(SoundCloudHTML5WidgetEvents.PLAY);
    this.soundcloudPlayer.unbind(SoundCloudHTML5WidgetEvents.PLAY_PROGRESS);
    this.soundcloudPlayer.unbind(SoundCloudHTML5WidgetEvents.ERROR);
    this.soundcloudPlayer.unbind(SoundCloudHTML5WidgetEvents.READY);
  }

  onReady = (): void => {
    if (!this.soundcloudPlayer) {
      return;
    }
    const { onTrackEnd } = this.props;
    this.soundcloudPlayer.bind(SoundCloudHTML5WidgetEvents.FINISH, onTrackEnd);
    this.soundcloudPlayer.bind(SoundCloudHTML5WidgetEvents.PAUSE, this.onPause);
    this.soundcloudPlayer.bind(SoundCloudHTML5WidgetEvents.PLAY, this.onPlay);
    this.soundcloudPlayer.bind(
      SoundCloudHTML5WidgetEvents.PLAY_PROGRESS,
      _throttle(this.onProgressChange, 2000, { leading: true, trailing: true })
    );
    this.soundcloudPlayer.bind(SoundCloudHTML5WidgetEvents.ERROR, this.onError);
  };

  onProgressChange = (event: ProgressEvent): void => {
    const { onProgressChange } = this.props;
    onProgressChange(event.currentPosition);
  };

  onPause = (event: ProgressEvent): void => {
    const { onPlayerPausedChange } = this.props;
    onPlayerPausedChange(true);
  };

  onPlay = (event: ProgressEvent): void => {
    const { onPlayerPausedChange } = this.props;
    // Detect new track loaded
    const { currentSoundId } = this.state;
    if (event.soundId !== currentSoundId) {
      this.setState({ currentSoundId: event.soundId });
      this.updateTrackInfo(event);
    }

    onPlayerPausedChange(false);
  };

  playListen = (listen: Listen | JSPFTrack) => {
    const { show, onTrackNotFound } = this.props;
    if (!show) {
      return;
    }
    const originURL = _get(listen, "track_metadata.additional_info.origin_url");
    if (/soundcloud\.com/.test(originURL)) {
      if (this.soundcloudPlayer) {
        this.soundcloudPlayer.load(originURL, this.options);
      } else if (this.retries <= 3) {
        this.retries += 1;
        setTimeout(this.playListen.bind(this, listen), 500);
      } else {
        // Abort!
        const { onInvalidateDataSource } = this.props;
        onInvalidateDataSource(
          this,
          "Soundcloud player did not load properly."
        );
      }
    } else {
      onTrackNotFound();
    }
  };

  updateTrackInfo = (event?: ProgressEvent) => {
    const {
      onTrackInfoChange,
      onDurationChange,
      onProgressChange,
    } = this.props;
    if (!this.soundcloudPlayer) {
      return;
    }
    this.soundcloudPlayer.getCurrentSound((currentTrack: any) => {
      if (!currentTrack) {
        return;
      }
      onTrackInfoChange(currentTrack.title, currentTrack.user?.username);
      onDurationChange(currentTrack.full_duration);
      if (event) {
        onProgressChange(event.currentPosition);
      }
    });
  };

  togglePlay = (): void => {
    if (!this.soundcloudPlayer) {
      return;
    }
    this.soundcloudPlayer.toggle();
  };

  seekToPositionMs = (msTimecode: number) => {
    if (!this.soundcloudPlayer) {
      return;
    }
    this.soundcloudPlayer.seekTo(msTimecode);
  };

  onError = (error: any): void => {
    const { handleError, onTrackNotFound } = this.props;
    handleError(error, "SoundCloud player error");
    onTrackNotFound();
  };

  render() {
    const { show } = this.props;
    return (
      <div className={`soundcloud ${!show ? "hidden" : ""}`}>
        <iframe
          id="soundcloud-iframe"
          ref={this.iFrameRef}
          title="Soundcloud player"
          width="100%"
          height="420px"
          scrolling="no"
          frameBorder="no"
          allow="autoplay"
          src="https://w.soundcloud.com/player/?url="
        />
      </div>
    );
  }
}
