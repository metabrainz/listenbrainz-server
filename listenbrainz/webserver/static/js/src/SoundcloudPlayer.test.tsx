import * as React from "react";
import { mount, shallow } from "enzyme";

import SoundcloudPlayer from "./SoundcloudPlayer";
import { DataSourceTypes } from "./BrainzPlayer";

const props = {
  show: true,
  playerPaused: false,
  onPlayerPausedChange: (paused: boolean) => {},
  onProgressChange: (progressMs: number) => {},
  onDurationChange: (durationMs: number) => {},
  onTrackInfoChange: (title: string, artist?: string) => {},
  onTrackEnd: () => {},
  onTrackNotFound: () => {},
  handleError: (error: BrainzPlayerError, title?: string) => {},
  handleWarning: (message: string | JSX.Element, title?: string) => {},
  handleSuccess: (message: string | JSX.Element, title?: string) => {},
  onInvalidateDataSource: (
    dataSource?: DataSourceTypes,
    message?: string | JSX.Element
  ) => {},
};

describe("SoundcloudPlayer", () => {
  it("renders", () => {
    window.fetch = jest.fn();
    const wrapper = mount(<SoundcloudPlayer {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should play if origin_url is a soundcloud URL", () => {
    const soundcloudListen: Listen = {
      listened_at: 42,
      track_metadata: {
        additional_info: {
          origin_url: "https://soundcloud.com/wankelmut/wankelmut-here-to-stay",
        },
        artist_name: "Wankelmut",
        release_name: "",
        track_name: "Rock'n'Roll Is Here To Stay",
      },
    };
    const onInvalidateDataSource = jest.fn();
    const onTrackNotFound = jest.fn();
    const mockProps = {
      ...props,
      onInvalidateDataSource,
      onTrackNotFound,
    };
    const wrapper = mount<SoundcloudPlayer>(
      <SoundcloudPlayer {...mockProps} />
    );

    const instance = wrapper.instance();
    if (!instance.soundcloudPlayer) {
      throw new Error("no SoundcloudPlayer");
    }
    instance.soundcloudPlayer.load = jest.fn();
    // play listen should extract the Soundcloud URL
    instance.playListen(soundcloudListen);
    expect(instance.soundcloudPlayer.load).toHaveBeenCalledTimes(1);
    expect(instance.soundcloudPlayer.load).toHaveBeenCalledWith(
      "https://soundcloud.com/wankelmut/wankelmut-here-to-stay",
      expect.any(Object)
    );
    expect(onTrackNotFound).not.toHaveBeenCalled();
    expect(onInvalidateDataSource).not.toHaveBeenCalled();
  });

  it("should ignore listens if origin_url is not a soundcloud URL", () => {
    const spotifyListen: Listen = {
      listened_at: 0,
      track_metadata: {
        artist_name: "Moondog",
        track_name: "Bird's Lament",
        additional_info: {
          spotify_id: "https://open.spotify.com/track/surprise!",
        },
      },
    };
    const onInvalidateDataSource = jest.fn();
    const onTrackNotFound = jest.fn();
    const mockProps = {
      ...props,
      onInvalidateDataSource,
      onTrackNotFound,
    };
    const wrapper = mount<SoundcloudPlayer>(
      <SoundcloudPlayer {...mockProps} />
    );
    const instance = wrapper.instance();

    if (!instance.soundcloudPlayer) {
      throw new Error("no SoundcloudPlayer");
    }
    instance.soundcloudPlayer.load = jest.fn();

    instance.playListen(spotifyListen);
    expect(onTrackNotFound).toHaveBeenCalledTimes(1);
    expect(onInvalidateDataSource).not.toHaveBeenCalled();
  });

  it("should update track info if playing a new track", () => {
    const onPlayerPausedChange = jest.fn();
    const onTrackInfoChange = jest.fn();
    const onDurationChange = jest.fn();
    const onProgressChange = jest.fn();
    const mockProps = {
      ...props,
      onPlayerPausedChange,
      onTrackInfoChange,
      onDurationChange,
      onProgressChange,
    };
    const wrapper = mount<SoundcloudPlayer>(
      <SoundcloudPlayer {...mockProps} />
    );
    const instance = wrapper.instance();
    instance.setState({ currentSoundId: 1 });
    if (!instance.soundcloudPlayer) {
      throw new Error("no SoundcloudPlayer");
    }
    instance.soundcloudPlayer.getCurrentSound = jest.fn((callback) =>
      callback({
        title: "Dope track",
        user: { username: "Emperor Norton the 1st" },
        full_duration: 420,
      })
    );

    instance.onPlay({
      soundId: 2,
      loadedProgress: 123,
      currentPosition: 456,
      relativePosition: 789,
    });
    expect(instance.state.currentSoundId).toEqual(2);
    expect(onPlayerPausedChange).toHaveBeenCalledTimes(1);
    expect(onPlayerPausedChange).toHaveBeenCalledWith(false);
    expect(onTrackInfoChange).toHaveBeenCalledWith(
      "Dope track",
      "Emperor Norton the 1st"
    );
    expect(onProgressChange).toHaveBeenCalledWith(456);
    expect(onDurationChange).toHaveBeenCalledWith(420);
  });

  it("should instruct the player to toggle play if togglePlay is called", () => {
    const wrapper = mount<SoundcloudPlayer>(<SoundcloudPlayer {...props} />);
    const instance = wrapper.instance();
    if (!instance.soundcloudPlayer) {
      throw new Error("no SoundcloudPlayer");
    }
    instance.soundcloudPlayer.toggle = jest.fn();

    instance.togglePlay();

    expect(instance.soundcloudPlayer.toggle).toHaveBeenCalledTimes(1);
  });

  it("should instruct the player seek to a position if seekToPositionMs is called", () => {
    const wrapper = mount<SoundcloudPlayer>(<SoundcloudPlayer {...props} />);
    const instance = wrapper.instance();
    if (!instance.soundcloudPlayer) {
      throw new Error("no SoundcloudPlayer");
    }
    instance.soundcloudPlayer.seekTo = jest.fn();

    instance.seekToPositionMs(1234);

    expect(instance.soundcloudPlayer.seekTo).toHaveBeenCalledTimes(1);
    expect(instance.soundcloudPlayer.seekTo).toHaveBeenCalledWith(1234);
  });
});
