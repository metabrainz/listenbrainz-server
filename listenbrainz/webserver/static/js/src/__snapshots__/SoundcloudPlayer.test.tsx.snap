// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SoundcloudPlayer renders 1`] = `
<SoundcloudPlayer
  handleError={[Function]}
  handleSuccess={[Function]}
  handleWarning={[Function]}
  onDurationChange={[Function]}
  onInvalidateDataSource={[Function]}
  onPlayerPausedChange={[Function]}
  onProgressChange={[Function]}
  onTrackEnd={[Function]}
  onTrackInfoChange={[Function]}
  onTrackNotFound={[Function]}
  playerPaused={false}
  show={true}
>
  <div
    className="soundcloud "
  >
    <iframe
      allow="autoplay"
      frameBorder="no"
      height="420px"
      id="soundcloud-iframe"
      scrolling="no"
      src="https://w.soundcloud.com/player/?url="
      title="Soundcloud player"
      width="100%"
    />
  </div>
</SoundcloudPlayer>
`;
